### DISCUSSION TITLE
UEFI Boot from dnsmasq via iPXE fails

### DISCUSSION THREAD

**Discussion by PadawanDeluxe** [ipxe/ipxe#474] [Q&A] [answered]
Hello folks,
I really appreciate your help to find my error made. 

I have an environment running the following components:
dnsmasq for DHCP 
TFTP server
NFS Server
Webserver (lighttpd)

Everything runs on a single OpenWRT router which shall store some tools for diagnostic,backup,restore,live-boot ect. So I have some test machine on my environment based on a Proxmox server. One is configured to boot BIOS the other one to boot UEFI. 

Alright.... first things first:

DNSMASQ Ruleset:
`dhcp-vendorclass=set:BIOS,PXEClient:Arch:00000`
`dhcp-vendorclass=set:UEFI32,PXEClient:Arch:00006`
`dhcp-vendorclass=set:UEFI,PXEClient:Arch:00007`
`dhcp-vendorclass=set:UEFI64,PXEClient:Arch:00009`
`dhcp-boot=tag:BIOS,http://myserver/boot/boot.ipxe`
`dhcp-boot=tag:UEFI32,http://myserver/boot/uefi32.efi`
`dhcp-boot=tag:UEFI,http://myserver/boot/uefi.efi`
`dhcp-boot=tag:UEFI64,http://myserver/boot/uefi64.efi`

My *.ipxe files will be accessible and I will be able to run iPXE from the http://boot.salstar.sk website, but only for the BIOS enabled machine. The UEFI machine will try to boot via PXE, but it will fail and return to the shell. Also if I configure 
'dhcp-boot=http://myserver/boot/' 

I will not be able to start the UEFI machine via iPXE. As I am currently learning how to deal with iPXE i decided to head over to you guys and ask what I am doing wrong. I consider my mistake is either a wrong configuration in dnsmasq or an incorrect configuration of the iPXE. 

To avoid mistakes I decided to use the [Robin Smidsrods files ](https://gist.github.com/robinsmidsrod/2234639) provided here on GitHub, but I still dont get a boot menu on the UEFI machine here.... 

Thanks a lot to everyone helping me out!

regards
Carsten

**Comment by NiKiZe** [comment]
From my experience with dnsmasq, add some more conditions (tags)
Please see https://ipxe.org/appnote/proxydhcp (same as https://gist.github.com/NiKiZe/5c181471b96ac37a069af0a76688944d)

The logic is to send the right options only when the tag matches. (I'm not saying to use that config, just to take ideas from it)
Let me know if you need more specific details. (you use `dhcp-boot` instead of `pxe-service`)

I would also ask that you try to cleanup the formatting, your config is a bit hard to read.

**Comment by PadawanDeluxe** [comment, marked as answer]
Thanks for your response. 

I have re-formatted the code in my original post so it should be more readable now. Sorry about that. I started over with the How-to and did a new ruleset with the following config:

`# Based on gist https://gist.github.com/NiKiZe/5c181471b96ac37a069af0a76688944d`
`# dhcp-match=set:<tag>,<option number>|option:<option name>|vi-encap:<enterprise>[,<value>]`
`# Based on logic in https://gist.github.com/robinsmidsrod/4008017`
`# iPXE sends a 175 option, checking suboptions`
`dhcp-match=set:ipxe-http,175,19`
`dhcp-match=set:ipxe-https,175,20`
`dhcp-match=set:ipxe-menu,175,39`
`dhcp-match=set:ipxe-pxe,175,33`
`dhcp-match=set:ipxe-bzimage,175,24`
`dhcp-match=set:ipxe-iscsi,175,17`
`dhcp-match=set:ipxe-efi,175,36`

`# set ipxe-ok tag if we have correct combination`
`# http && menu && iscsi ((pxe && bzimage) || efi)`
`tag-if=set:ipxe-ok,tag:ipxe-http,tag:ipxe-menu,tag:ipxe-iscsi,tag:ipxe-pxe,tag:ipxe-bzimage`
`tag-if=set:ipxe-ok,tag:ipxe-http,tag:ipxe-menu,tag:ipxe-iscsi,tag:ipxe-efi`

`#pxe-service=[tag:<tag>,]<CSA>,<menu text>[,<basename>|<bootservicetype>][,<server address>|<server_name>]`
`#pxe-prompt=[tag:<tag>,]<prompt>[,<timeout>]`
`# these create option 43 cruft, which is required in proxy mode`
`# TFTP IP is required on all dhcp-boot lines (unless dnsmasq itself acts as tftp server?)`

`pxe-service=tag:!ipxe-ok,X86-64_EFI,"PXEX8664EFI","ipxe/snponly.efi",1.1.1.1`
`pxe-service=tag:!ipxe-ok,X86-64_EFI,"PXEX8664EFI","ipxe/snponly.efi",1.1.1.1`

`# later match overrides previous, keep ipxe script last`
`# server address must be non zero, but can be anythin as long as iPXE script is not fetched over TFTP`
`#dhcp-boot=[tag:<tag>,]<filename>,[<servername>[,<server address>|<tftp_servername>]]`
`pxe-prompt=tag:ipxe-ok,"Going to boot iPXE Environment...",0`
`dhcp-boot=tag:ipxe-ok,http://sngw02.stamm-networks.lan/boot/boot.ipxe`
`# OR Salstar Online Bootserver`
`#dhcp-boot=tag:ipxe-ok,http://boot.salstar.sk/`


For all those who are struggeling with UEFI:
**You have to define the entry for booting to the UEFI file twice - no matter what the manual says.** I found the hint regarding this on a german website. Credits goes [here](https://ssl.hehoe.de/artikel/blog/dnsmasq-fur-pxe-boot.html) So finally I ended up to be able to run whatever EFI rom I want to load. 

I can stand, that this is affecting dnsmasq on OpenWRT maybe someone can test it on a Raspberry or something similar? Anyhow thanks to community for help!


regards
Carsten

**Reply by NiKiZe** [reply]
please use proper code blocks using ``` instead.

Since you are running a full DHCP (at least in your question) I would recommend that you skip the PXE prompt and services, that just makes life more difficult ;)

And it makes no sense to give efi multiple times at all.
This config has no support for pcbios ?

**Reply by PadawanDeluxe** [reply]
I made use of the code blocks in the first post, but somehow formatting was destroyed.  That's why it looks so terrible maybeI should go for a tut here. 

I have used the Prompt command to debug my start. That helped me a lot to see whether the right code block was used or not. 

I do not understand what you mean with the missing pcbios? I consider, that I need to add this:

`pxe-service=tag:!ipxe-ok,X86PC,"PXEX86PC",<IMAGENAME>,1.1.1.1`

So it will go to the tftp and start the PXELinux or whatever I want. Am I correct with this assumption?
